src/mgr/ActivePyModules.cc

```diff
@@ -538,6 +538,14 @@ void ActivePyModules::start_one(PyModuleRef py_module)
   // Send all python calls down a Finisher to avoid blocking
   // C++ code, and avoid any potential lock cycles.
   finisher.queue(new LambdaContext([this, active_module, name](int) {
```

Comment:
So an issue I see here is that this module failing to load is presumably starving this finisher. We now have a per-module finisher so this may not actually block the completion of most work but, I just checked, we have code like `DaemonServer::handle_osd_perf_metric_query_updated` which does queue contexts on this finisher. If a module is legit stuck loading indefinitely then those finishers will never run. Correct? We also can see that we have a pipeline of modules loading and one becoming stuck blocks all subsequent modules, no?

So that means we need a way to terminate the module load (probably not trivial) or launch a thread for each module load to avoid blocking this Finisher thread. My preference would be to just launch a thread which loads the module then exits.

Reply:
Well, looking at this code again. I see that there is `class ActivePyModule : public PyModuleRunner` which is created just before this code above (`active_module`). That `PyModuleRunner` actually creates a thread already just for calling `->serve()` in the module. Well, we could just have that thread first `load` and only then `serve`. If you attach a context to the PyModuleRunner then you can have it complete the context after `load` completes. That would allow you to execute this lambda above.

Anyway, I think that's the most elegant way to resolve this.

Reply:
ACK, I created a ticket here: https://tracker.ceph.com/issues/73060